pkg/strategy/grid2/strategy.go

```diff
@@ -1438,6 +1487,78 @@ func (s *Strategy) buildPinOrderMap(grid *Grid, openOrders []types.Order) (map[s
 	return pinOrderMap, nil
 }
 
+// buildFilledPinOrderMapFromTrades will query the trades from last 24 hour and use them to build a pin order map
+// It will skip the orders on pins at which open orders are already
+func (s *Strategy) buildFilledPinOrderMapFromTrades(ctx context.Context, historyService types.ExchangeTradeHistoryService, openPinOrderMap PinOrderMap) (PinOrderMap, error) {
+	filledPinOrderMap := make(PinOrderMap)
+
+	var limit int64 = 1000
+	// get the filled orders when bbgo is down in order from trades
+	// [NOTE] only retrieve from last 24 hours !!!
+	var fromTradeID uint64 = 0
+	for {
+		trades, err := historyService.QueryTrades(ctx, s.Symbol, &types.TradeQueryOptions{
+			LastTradeID: fromTradeID,
+			Limit:       limit,
+		})
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to query trades to recover the grid with open orders")
+		}
+
+		s.logger.Infof("[DEBUG] len of trades: %d", len(trades))
+
+		for _, trade := range trades {
+			order, err := s.orderQueryService.QueryOrder(ctx, types.OrderQuery{
+				OrderID: strconv.FormatUint(trade.OrderID, 10),
+			})
```

Comment:
You can create a cache to store queried orders because multiple trades can be mapped to the same order id. The current implementation can cause duplicate queries.

Reply:
Also, to further speed up, if an order id is found in the open orders, you don't have to query it again and can skip current iteration as well.

Reply:
yes, this might cause duplicated order queries